src/formatter.ts

```diff
@@ -404,48 +417,137 @@ export class Formatter {
     return this.minTotalWidth;
   }
 
+  // calculates the resolution multiplier for `voices`.
+  static getResolutionMultiplier(voices: Voice[]): number {
+    if (!voices || !voices.length) {
+      throw new Vex.RERR('BadArgument', 'No voices to format');
+    }
+    const totalTicks = voices[0].getTotalTicks();
+    const resolutionMultiplier = voices.reduce((accumulator, voice) => {
+      if (!voice.getTotalTicks().equals(totalTicks)) {
+        throw new Vex.RERR('TickMismatch', 'Voices should have same total note duration in ticks.');
+      }
+
+      if (voice.getMode() === Voice.Mode.STRICT && !voice.isComplete()) {
+        throw new Vex.RERR('IncompleteVoice', 'Voice does not have enough notes.');
+      }
+
+      return Math.max(accumulator, Fraction.LCM(accumulator, voice.getResolutionMultiplier()));
+    }, 1);
+    return resolutionMultiplier;
+  }
+
   // Create `ModifierContext`s for each tick in `voices`.
-  createModifierContexts(voices) {
-    const contexts = createContexts(voices, ModifierContext, (tickable, context) =>
-      tickable.addToModifierContext(context)
-    );
+  createModifierContexts(voices: Voice[]): ModifierContextsStruct {
```

Comment:
Can you share the logic with a template method?

```typescript
function createContexts<T>(voices: Voice[]) : FormatterContexts<T> {
   ...
}
```

Reply:
I will give it a try.

Reply:
The compiler complains in three places:
 `const newContext = new T({ tickID: integerTicks });`it seems that new is not allowed with T. There is also a different number of parameters on both constructors.
`tickToContextMap[integerTicks].addTickable(tickable, voiceIndex);` T does not have this method
```
contexts.forEach((context) => {
      context.tContexts = contexts;
    });
```

Reply:
Right, seems like TS doesn't like that, and you'll need to pass a builder parameter. I was able to get this prototype to work on the TypeScript playground (see the code after HERE):

```typescript

// Mock class for playground
class ModifierContext {}

// Mock class for playground
class TickContext {
    constructor(tick?: {tickID: number}) {}
    addTickable(n: number): void {}
}

// Mock class for playground
class Tickable {
    addToModifierContext(c: ModifierContext): void {}
}

// New formatter prototype begins HERE:

interface AlignmentContexts<T> {
    map: Record<number, T>,
    array: T[],
    list: T[]
    resolutionMultiplier: number
}

type addToContextFn<T> = (tickable: Tickable, context: T, voiceIndex?: number) => void
type makeContextFn<T> = (tick?: {tickID: number}) => T

function createContexts<T>(voices: string[], makeContext: makeContextFn<T>, addToContext: addToContextFn<T>): AlignmentContexts<T> {
    const c = makeContext();
    return {
        map: {},
        array: [],
        list: [],
        resolutionMultiplier: 1,
    };
}

function createModifierContexts(voices: string[]): AlignmentContexts<ModifierContext> {
    const fn: addToContextFn<ModifierContext> = (tickable: Tickable, context: ModifierContext, voiceIndex?: number) => tickable.addToModifierContext(context);
    return createContexts(voices, () => new ModifierContext(), fn);
}

function createTickContexts(voices: string[]): AlignmentContexts<TickContext> {
    const fn: addToContextFn<TickContext> = (tickable: Tickable, context: TickContext, voiceIndex?: number) => (context as TickContext).addTickable(voiceIndex as number);
    return createContexts(voices, (tick?: {tickID: number}) => new TickContext(tick), fn);
}
```

Reply:
done